lxd/db/query/slices_test.go

```diff
@@ -0,0 +1,73 @@
+package query_test
+
+import (
+	"database/sql"
+	"fmt"
+	"testing"
+
+	_ "github.com/mattn/go-sqlite3"
+	"github.com/mpvl/subtest"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/lxc/lxd/lxd/db/query"
+)
+
+// Exercise possible failure modes.
+func TestStrings_Error(t *testing.T) {
+	for _, c := range testStringsErrorCases {
+		subtest.Run(t, c.query, func(t *testing.T) {
+			tx := newTxForSlices()
+			values, err := query.Strings(tx, c.query)
+			assert.EqualError(t, err, c.error)
+			assert.Nil(t, values)
+		})
+	}
+}
+
+// All values yield by the query are returned.
+func TestStrings(t *testing.T) {
+	tx := newTxForSlices()
+	values, err := query.Strings(tx, "SELECT name FROM test ORDER BY name")
+	assert.Nil(t, err)
+	assert.Equal(t, []string{"bar", "foo"}, values)
+}
+
+// Exercise possible failure modes.
+func TestIntegers_Error(t *testing.T) {
+	for _, c := range testIntegersErrorCases {
+		subtest.Run(t, c.query, func(t *testing.T) {
+			tx := newTxForSlices()
+			values, err := query.Integers(tx, c.query)
+			assert.EqualError(t, err, c.error)
+			assert.Nil(t, values)
+		})
+	}
+}
+
+// All values yield by the query are returned.
+func TestIntegers(t *testing.T) {
+	tx := newTxForSlices()
+	values, err := query.Integers(tx, "SELECT id FROM test ORDER BY id")
+	assert.Nil(t, err)
+	assert.Equal(t, []int{0, 1}, values)
+}
+
+// Return a new transaction against an in-memory SQLite database with a single
+// test table populated with a few rows.
+func newTxForSlices() *sql.Tx {
+	db, err := sql.Open("sqlite3", ":memory:")
+	if err != nil {
+		panic(fmt.Sprintf("failed to create database: %v", err))
+	}
+	if _, err := db.Exec("CREATE TABLE test (id INTEGER, name TEXT)"); err != nil {
+		panic(fmt.Sprintf("failed to create table: %v", err))
+	}
+	if _, err := db.Exec("INSERT INTO test VALUES (0, 'foo'), (1, 'bar')"); err != nil {
+		panic(fmt.Sprintf("failed to insert rows: %v", err))
+	}
+	tx, err := db.Begin()
+	if err != nil {
+		panic(fmt.Sprintf("failed to begin transaction: %v", err))
+	}
+	return tx
```

Comment:
Yes, AFAIU using panic in this context is the "right thing", modulo adopting Go's mindset and established idioms.

For example [this](https://pastebin.com/r8k7wgsH) is the result of grep-ing "panic(" across all *_test.go files in the standard library, e.g.:

```
func newCipher(key []byte) *desCipher {
	c, err := NewCipher(key)
	if err != nil {
		panic("NewCipher failed: " + err.Error())
	}
	return c.(*desCipher)
}
```

The rational for it is in Go tests are considered to be "just a program", subject to the same programming style as normal code (and that's one of the reason the stdlib testing package is so bare bones). In a normal program if you can't handle an error you either return it up the call stack (which might know how how to handle it) or panic. In this case it doesn't make much sense to return the error value up, since we know already we are going to bail out, as no consumer of these test helper would be coded to handle such errors.

Also, if one passed the error up to the TestXXX function being executed and then called t.Fatal(err), that would result in somehow suboptimal failure output, because you would not know where the error originated, while with panic() you get the traceback. BTW this the reinforces the idea that test code is just regular Go code, as for good and for bad this could be considered a problem also for "normal" Go programs, or at least is one of the critics that Go's error handling patterns typically receive. This point is not exactly an argument in favor of using panic() in cases like this, but it shows the rational.

All this being said, the assert.* functions would surely work, but they require passing the *testing.T reference around (in this specific case it would be an extra argument of the newTxForSlices function), and when testing setup becomes more complex there might be a lot of passing around to be done (e.g. helpers of helpers). But more importantly, if you use assert.*, when a test fails because of setup issues, you don't get a stack trace.

I'm not particularly strong either way, if you prefer assert.*, that'd be fine for me.

FWIW I don't think using panic() in test setup functions makes possible uses of panic() in actual LXD non-test code any more or less legitimate, strictly speaking. I do however understand it might feel as such to somebody not particularly familiar with Go, but I'm not sure it's really a concern we should care about, as a similar point would hold with many other aspects, and it's not clear where one should draw the line (i.e. "I do not use this specific idiomatic pattern because casual contributors unfamiliar with the language might not understand how to use it properly"). I guess external contributions need to be double/tripled checked and "sanitized" anyways.